elastic/apm-server
# Move source mapping to model processing

## Motivation/summary

Source mapping transformation is rewritten as a `model.BatchProcessor`, moved out of the model package and into the sourcemap package. Moving such logic out of the model package takes us another step towards generating the model types and `beat.Event` transformation.

One side-effect of the move to a process, worth debating, is that source mapping now occurs in the HTTP handler goroutine rather than in the publisher, which could increase request latency slightly. On the other hand this means that source mapping for more clients than there are CPUs can now happen concurrently (the publisher is limited to the number of CPUs in this regard); and the handlers will block for up to one second anyway if the publisher is busy/queue is full.

If a stack frame cannot be mapped, we no longer set `sourcemap.error` or log anything. Just because RUM and source mapping is enabled, does not mean that all stacks _must_ be source mapped; therefore these "errors" are mostly just noise. Likewise we now only set `sourcemap.updated` when it is true.

I removed the `TestSourcemapCacheExpiration` system test as I believe it is redundant. In the test immediately above that one we prove that caching is effective, and cache expiration is unit tested.

## Checklist

~- [ ] Documentation has been updated~

## How to test these changes

1. Enable RUM, upload a source map, check that the source map is applied to indexed events
2. Enable RUM, _don't_ upload a source map, check that events are indexed without `sourcemap.error` or `sourcemap.updated `set

## Related issues

Prerequisite for https://github.com/elastic/apm-server/issues/4120 and https://github.com/elastic/apm-server/issues/3565
Closes https://github.com/elastic/apm-server/issues/4958
